crowbar_framework/app/models/api/node.rb

```diff
@@ -26,67 +26,41 @@ def upgraded?
       @upgraded ||= current_platform == @node[:target_platform]
     end
 
-    def pre_upgrade
-      # TODO: save the global status info about this substep (we started prepare for upgrade)
-
-      # Migrate out the l3 agents and shut down pacemaker
-      script = "/usr/sbin/crowbar-pre-upgrade.sh"
-      ssh_status = @node.ssh_cmd(script)
-      if ssh_status[0] != 200
-        save_error_state("Executing of pre upgrade script has failed on node #{@node.name}.")
-        return false
-      end
+    # execute script in background and wait for it to finish
+    def execute_and_wait_for_finish(script, seconds)
+      Rails.logger.info("Executing #{script} at #{@node.name}...")
+      @node.wait_for_script_to_finish(script, seconds)
+      true
+    rescue StandardError => e
+      save_error_state(
+        e.message + "Check /var/log/crowbar/node-upgrade.log for details."
+      )
+      false
+    end
 
-      upgrade_ok = false
-      begin
-        Timeout.timeout(60) do
-          loop do
-            out = @node.run_ssh_cmd("test -e /var/lib/crowbar/upgrade/crowbar-pre-upgrade-ok")
-            if out[:exit_code].zero?
-              save_node_state("Pre upgrade script run was successful.")
-              upgrade_ok = true
-              break
-            end
-            out = @node.run_ssh_cmd("test -e /var/lib/crowbar/upgrade/crowbar-pre-upgrade-failed")
-            if out[:exit_code].zero?
-              save_error_state("Execution of pre upgrade script at node #{@node.name} has failed
-Check /var/log/crowbar/node-upgrade.log at #{@node.name} node for details.")
-              break
-            end
-            sleep(5)
-          end
-        end
-      rescue Timeout::Error
-        save_error_state("Possible error during pre-upgrade script execution.
-Action did not finish after 1 minute")
-        return false
+    def pre_upgrade
+      if execute_and_wait_for_finish("/usr/sbin/crowbar-pre-upgrade.sh", 300)
+        save_node_state("Pre upgrade script run was successful.")
```

Comment:
Not really. All of these actions are part of `nodes` step, but they are different sub-steps. We haven't defined API for sub-steps yet, so there's only this preliminary `save_node_state` method which does not do anything so far